lib/ts/recipe/emailpassword/api/implementation.ts

```diff
@@ -586,66 +597,272 @@ export default function getAPIImplementation(): APIInterface {
+                            tenantId,
+                            userContext,
+                        });
+
+                        if (signInResponse.status === "WRONG_CREDENTIALS_ERROR") {
+                            throw new SignInError(signInResponse);
+                        }
+
+                        await assertThatSignInIsAllowed(tenantId, signInResponse.user, userContext);
+
+                        if (!(await isValidFirstFactor(tenantId, "emailpassword", userContext))) {
+                            throw new SessionError({
+                                type: SessionError.UNAUTHORISED,
+                                message: "Session is required for secondary factors",
+                                payload: {
+                                    clearTokens: false,
+                                },
+                            });
+                        }
+
+                        signInResponse.user = await await AccountLinking.getInstance().createPrimaryUserIdOrLinkAccounts(
+                            {
+                                tenantId,
+                                user: signInResponse.user,
+                                userContext,
+                            }
+                        );
+
+                        session = await Session.createNewSession(
+                            options.req,
+                            options.res,
+                            tenantId,
+                            signInResponse.recipeUserId,
+                            {},
+                            {},
+                            userContext
+                        );
+
+                        await MultiFactorAuth.markFactorAsCompleteInSession(session, "emailpassword", userContext);
+
+                        return {
+                            status: "OK",
+                            session,
+                            user: signInResponse.user,
+                        };
+                    } else {
+                        // This branch - MFA is enabled / Active Session (secondary factor) / Sign in
+
+                        let sessionUser = await getUser(session.getUserId(), userContext);
+                        if (sessionUser === undefined) {
+                            throw new SessionError({
+                                type: SessionError.UNAUTHORISED,
+                                message: "Session user not found",
+                            });
+                        }
+
+                        let signInResponse = await options.recipeImplementation.signIn({
+                            email,
+                            password,
+                            tenantId,
+                            userContext,
+                        });
+
+                        if (signInResponse.status === "WRONG_CREDENTIALS_ERROR") {
+                            throw new SignInError(signInResponse);
+                        }
+
```

Comment:
add check for assetSignInAllowed and mention that it will always pass cause of <REASON>, however we add it just in case the implementation of the function changes in the future